### PR TITLE
Add sharing exceptions to OCP

### PR DESCRIPTION
**Pull request by rullzer**
The title says it all.

CC: @schiesbn @PVince81 @nickvergessen

### CODE REVIEW

File: lib/private/share20/defaultshareprovider.php [hunk 1]
```diff
@@ -24,7 +24,7 @@
 use OCP\Share\IShareProvider;
 use OC\Share20\Exception\InvalidShare;
 use OC\Share20\Exception\ProviderException;
-use OC\Share20\Exception\ShareNotFound;
+use OCP\Share\Exceptions\ShareNotFound;
```
Comment (nickvergessen) [inline lib/private/share20/defaultshareprovider.php]: the autism in me explodes here

Comment (rullzer) [inline lib/private/share20/defaultshareprovider.php]: hehe. Ah yes I saw that encryption uses plurar in OCP... so I just followed that standard...

File: lib/public/share/exceptions/genericshareexception.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	/**
+	 * @param string $message
+	 * @param string $hint
+	 * @param int $code
+	 * @param \Exception $previous
+	 * @since 9.0.0
+	 */
+	public function __construct($message = '', $hint = '', $code = 0, \Exception $previous = null) {
+		if (empty($message)) {
+			$message = 'Unspecified share exception';
```
Comment (nickvergessen) [inline lib/public/share/exceptions/genericshareexception.php]: Should fallback to the hint first, if it is not empty?

**Comment by nickvergessen** [comment]
Looks good otherwise

File: lib/public/share/exceptions/genericshareexception.php (hunk 2 above)
Comment (rullzer) [inline lib/public/share/exceptions/genericshareexception.php]: mmmm i'm not sure... why would anybody do `new GenericShareException('', 'fix it!')`

Plus the hint does not have to tell you what went wrong.

Comment (nickvergessen) [inline lib/public/share/exceptions/genericshareexception.php]: I got confused with https://github.com/owncloud/core/pull/21942 but that is the other way around. Fine by me

Comment (schiesbn) [inline lib/public/share/exceptions/genericshareexception.php]: The hint is meant to be a translated string, used as output for the user. In most cases it doesn't make sense to use it as message, write it to the logfile, etc.

**Comment by schiesbn** [comment]
looks good to me :+1:

**Pull request merged**